### PR TITLE
Transform Main.hs into Tests

### PR DESCRIPTION
**Pull request by voidus**
Hi :wave: 

I started converting the [Main.hs](https://github.com/haskell-servant/servant-uverb/blob/4c5d20cdbbea585f69e92351a5264a96c7c62b33/example/Main.hs) from the [checklist in the main PR](https://github.com/haskell-servant/servant/pull/1314) into tests. I don't have a lot of experience with larger Haskell projects, so any feedback is appreciated.

### CODE REVIEW

**Comment by voidus** [comment]
My understanding is that this can only be reasonably tested in the servant-server and servant-client packages. I was peeking at type-spec and tao, but I don't think this would be the place to introduce them, even if there was consensus on that.

For now this is only the server part, I'll look into the client part as soon as I find time.

CONTRIBUTING.md mentioned -Werror-compatibility to check ghc 7.8 and 7.10 compatibility. I searched the web, but couldn't figure out what it means. I ran stylish and hlint though.

**Comment by voidus** [comment]
I think I found an issue with the current implementation. servant-client interprets responses with status 301 as an error and gives us a `Left ClientError` even if we have that response defined in the API, which isn't what I would expect.

I added a test for the current behaviour and a comment, since I think changing that is out of scope for this PR.

**Comment by voidus** [comment]
BTW, the tests for servant-http-streams seem to be broken or need some more setup that I can't figure out :neutral_face:

**Comment by voidus** [comment]
@fisx highlighting you since you seem to have done most work on the uverb branch.

**Comment by fisx** [comment]
thanks!  i'll take a look at this soon, very happy to see that you're picking this up!  :)

File: servant-client/test/Servant/ClientTestUtils.hs [hunk 1]
```diff
@@ -166,7 +202,12 @@ server = serve api (
   :<|> (return $ addHeader 1729 $ addHeader "eg2" True)
   :<|> return NoContent
   :<|> (Tagged $ \ _request respond -> respond $ Wai.responseLBS HTTP.found302 [("Location", "testlocation"), ("Set-Cookie", "testcookie=test")] "")
-  :<|> emptyServer)
+  :<|> emptyServer
+  :<|> (\shouldRedirect -> if shouldRedirect
+                              then respond (WithStatus @301 ("redirecting" :: Text))
+                              else respond (WithStatus @200 alice ))
```
Comment (fisx) [inline servant-client/test/Servant/ClientTestUtils.hs:208]:
```suggestion
                              else respond (WithStatus @200 alice))
```

File: servant-client/test/Servant/SuccessSpec.hs [hunk 2]
```diff
@@ -151,3 +157,44 @@ successSpec = beforeAll (startWaiApp server) $ afterAll endWaiApp $ do
...
+        --   case eitherResponse of
+        --     Left clientError -> fail $ show clientError
+        --     Right response ->
+        --       case response of
+        --         Z (Identity (WithStatus person :: WithStatus 200 Person)) ->
+        --           fail "Expected to be redirected"
+        --         S (last) ->
+        --           let Identity (WithStatus message :: WithStatus 301 Text)
+        --                 = unZ last
+        --           in message `shouldBe` "redirecting"
```
Comment (fisx) [inline servant-client/test/Servant/SuccessSpec.hs:176]:
```suggestion
        --     Right (Z (Identity (WithStatus person :: WithStatus 200 Person)) ->
        --       fail "Expected to be redirected"
        --     Right (S last) ->
        --       let Identity (WithStatus message :: WithStatus 301 Text) = unZ last
        --       in message `shouldBe` "redirecting"
```

(you could collapse the nexted cases in the actual test as well if you like, but i agree, that code shouldn't make it out of the uverb-branch.)

**Review by fisx — APPROVED** [review]
I left some nit-picks, but I'll merge this anyway because it's a step towards checking another box in the main PR.  Up to you if you want to open another PR.

[I agree](https://github.com/zerobuzz/servant/pull/1#issuecomment-648784387) and left this as a new check-box in the main PR.

I also agree that there isn't much to test in the `servant` package.

I couldn't find anything on `-Werror-compatible` either, so I suppose as long as we're passing the CI it's fine.  :)

**Comment by fisx** [comment]
Thanks again!

**Pull request merged**